include/vcl/Image.h

```diff
@@ -325,6 +326,432 @@ namespace VCL {
+        friend class ImageTest;
+
+    /*  *********************** */
+    /*        VARIABLES         */
+    /*  *********************** */
+        // Image height and width
+        uint _height, _width;
+
+        // Type of image (OpenCV definition) and number of channels
+        int _cv_type, _channels;
+
+        // Maintains order of operations requested
+        std::vector<std::shared_ptr<Operation>> _operations;
+
+        // Image format and compression type
+        Format _format;
+        CompressionType _compress;
+
+        // Full path to image
+        std::string _image_id;
+
+        // Image data (OpenCV Mat or TDBImage)
+        cv::Mat _cv_img;
+        TDBImage *_tdb;
+
+    /*  *********************** */
+    /*      COPY FUNCTIONS      */
+    /*  *********************** */
+        /**
+         *  Copies (deep copy) an OpenCV Mat into the Image OpenCV Mat
+         *
+         *  @param cv_img  An existing OpenCV Mat
+         */
+        void deep_copy_cv(const cv::Mat &cv_img);
+
+        /**
+         *  Copies (shallow copy) an OpenCV Mat into the Image OpenCV Mat
+         *
+         *  @param cv_img  An existing OpenCV Mat
+         */
+        void shallow_copy_cv(const cv::Mat &cv_img);
+
+        /**
+         *  Copies the Image OpenCV Mat into a buffer
+         *
+         *  @param buffer  The buffer that will contain the image
+         *    data
+         */
+        template <class T> void copy_to_buffer(T* buffer);
+
+    /*  *********************** */
+    /*      UTIL FUNCTIONS      */
+    /*  *********************** */
+
+        /**
+         *  Performs the set of operations that have been requested
+         *    on the Image
+         */
+        void perform_operations();
+
+        std::string format_to_string(Image::Format format);
+
+        /**
+         *  Creates full path to Image with appropriate extension based
+         *    on the Image::Format
+         *
+         *  @param filename The path to the Image object
+         *  @param format  The Image::Format of the Image object
+         *  @return Full path to the object including extension
+         */
+        std::string create_fullpath(const std::string &filename,
+            Image::Format format);
+
+    /*  *********************** */
+    /*        OPERATION         */
+    /*  *********************** */
+
+        enum class OperationType {
+            READ,
+            WRITE,
+            RESIZE,
+            CROP,
+            THRESHOLD,
+            FLIP,
+            ROTATE
+        };
+
+        /**
+         *  Provides a way to keep track of what operations should
+         *   be performed on the data when it is needed
+         *
+         *  Operation is the base class, it keeps track of the format
+         *   of the image data, defines a way to convert Format to
+         *   a string, and defines a virtual function that overloads the
+         *   () operator
+         */
+        class Operation {
```

Comment:
Since this commit is about simplifying code and removing duplicates, I see a lot of similarities between Operations in images and videos if that were to be a template with specialization. 
In any case, this header is too big and you might consider segregating operations part from the rest. We did that at some point with PMGDIterators since it was getting so complex to contain in one. The class was still a subclass but declarations and definitions were in separate header/cc files.

Reply:
I agree that a template for operations that receive the media (image, video) as type would look good. I suppose given the difficulty in reviewing this huge change set (due to removed/copied lines), we may be better of doing it in another patch.

As a side note, we discussed this as part of a more recent change. The use case is as follows: a frame is extracted from a video as result of 'FindFrame' command, and the command contains the operations of 'Image' type rather than 'Video'. Since the video command class processes and checks (in VideoCommand.c) only video operations, we had to expose Image operations through a dirty hack (as you'll see when reviewing FindFrames patch). Seems like templated operations may solve this very problem.

Reply:
Separate patch after this merge is fine. It will be nicely contained and still be part of the code cleanup sort of theme.